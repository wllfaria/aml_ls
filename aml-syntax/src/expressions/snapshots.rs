use aml_core::Location;
use aml_token::{Operator, Primitive, TokenKind};
use serde::Serialize;

use crate::ast::*;
use crate::parser::snapshots::ToSnapshot;

impl<'ast> ToSnapshot<'ast> for Expr {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        match self {
            Expr::Ident(location) => SnapshotExpr::Ident(SnapshotIdentifier {
                location,
                value: &content[location.to_range()],
            }),
            Expr::String(location) => SnapshotExpr::String(SnapshotString {
                location,
                value: &content[location.to_range()],
            }),
            Expr::Unary(unary) => unary.into_snapshot(content),
            Expr::Binary(binary) => binary.into_snapshot(content),
            Expr::Call(call) => call.into_snapshot(content),
            Expr::Primitive(primitive) => primitive.into_snapshot(content),
            Expr::ArrayIndex(array_index) => array_index.into_snapshot(content),
            Expr::List(list) => list.into_snapshot(content),
            Expr::Map(map) => map.into_snapshot(content),
            Expr::Error(error) => error.into_snapshot(content),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotUnary<'ast> {
    pub original: &'ast str,
    pub op: Operator,
    pub expr: Box<SnapshotExpr<'ast>>,
    pub location: Location,
}

impl<'ast> ToSnapshot<'ast> for Unary {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Unary(SnapshotUnary {
            op: self.op,
            expr: Box::new(self.expr.into_snapshot(content)),
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotBinary<'ast> {
    pub lhs: Box<SnapshotExpr<'ast>>,
    pub rhs: Box<SnapshotExpr<'ast>>,
    pub op: Operator,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for Binary {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Binary(SnapshotBinary {
            lhs: Box::new(self.lhs.into_snapshot(content)),
            rhs: Box::new(self.rhs.into_snapshot(content)),
            op: self.op,
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotIdentifier<'ast> {
    pub value: &'ast str,
    pub location: Location,
}

impl<'ast> ToSnapshot<'ast> for SnapshotIdentifier<'ast> {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Ident(SnapshotIdentifier {
            location: self.location,
            value: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotString<'ast> {
    pub value: &'ast str,
    pub location: Location,
}

impl<'ast> ToSnapshot<'ast> for SnapshotString<'ast> {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::String(SnapshotString {
            location: self.location,
            value: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotCall<'ast> {
    pub fun: Box<SnapshotExpr<'ast>>,
    pub args: Vec<SnapshotExpr<'ast>>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for Call {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Call(SnapshotCall {
            fun: Box::new(self.fun.into_snapshot(content)),
            args: self
                .args
                .into_iter()
                .map(|e| e.into_snapshot(content))
                .collect(),
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotPrimitive<'ast> {
    pub value: Primitive,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for PrimitiveExpr {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Primitive(SnapshotPrimitive {
            value: self.value,
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotArrayIndex<'ast> {
    pub lhs: Box<SnapshotExpr<'ast>>,
    pub index: Box<SnapshotExpr<'ast>>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for ArrayIndex {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::ArrayIndex(SnapshotArrayIndex {
            location: self.location,
            original: &content[self.location.to_range()],
            lhs: Box::new(self.lhs.into_snapshot(content)),
            index: Box::new(self.index.into_snapshot(content)),
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotList<'ast> {
    pub items: Vec<SnapshotExpr<'ast>>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for List {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::List(SnapshotList {
            location: self.location,
            original: &content[self.location.to_range()],
            items: self
                .items
                .into_iter()
                .map(|e| e.into_snapshot(content))
                .collect(),
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotMap<'ast> {
    pub items: Vec<(SnapshotExpr<'ast>, SnapshotExpr<'ast>)>,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for Map {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Map(SnapshotMap {
            location: self.location,
            original: &content[self.location.to_range()],
            items: self
                .items
                .into_iter()
                .map(|(l, r)| (l.into_snapshot(content), r.into_snapshot(content)))
                .collect(),
        })
    }
}

#[derive(Debug, Serialize)]
pub struct SnapshotErrorExpr<'ast> {
    pub token: TokenKind,
    pub location: Location,
    pub original: &'ast str,
}

impl<'ast> ToSnapshot<'ast> for ErrorExpr {
    type Item = SnapshotExpr<'ast>;

    fn into_snapshot(self, content: &'ast str) -> Self::Item {
        SnapshotExpr::Error(SnapshotErrorExpr {
            token: self.token,
            location: self.location,
            original: &content[self.location.to_range()],
        })
    }
}

#[derive(Debug, Serialize)]
pub enum SnapshotExpr<'ast> {
    Unary(SnapshotUnary<'ast>),
    Binary(SnapshotBinary<'ast>),
    Ident(SnapshotIdentifier<'ast>),
    String(SnapshotString<'ast>),
    Call(SnapshotCall<'ast>),
    Primitive(SnapshotPrimitive<'ast>),
    ArrayIndex(SnapshotArrayIndex<'ast>),
    List(SnapshotList<'ast>),
    Map(SnapshotMap<'ast>),
    Error(SnapshotErrorExpr<'ast>),
}
