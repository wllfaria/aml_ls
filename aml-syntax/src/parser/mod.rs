use aml_token::{Element, Operator, Token, TokenKind, Tokens};

use crate::ast::*;
use crate::expressions::parse_expression;

#[cfg(test)]
mod tests;

#[cfg(test)]
pub(crate) mod snapshots;

struct LocationCalculator;

impl LocationCalculator {
    fn merge_location_with_values(
        start: aml_core::Location,
        attributes: &Attributes,
        values: Option<&Vec<AstNode>>,
        children: Option<&Vec<AstNode>>,
    ) -> aml_core::Location {
        let value_location = values.and_then(|v| v.iter().last().map(|node| node.location()));
        let children_location = children.and_then(|c| c.iter().last().map(|node| node.location()));

        match (children_location, value_location, attributes.location) {
            (Some(location), _, _) => start.merge(location),
            (_, Some(location), _) => start.merge(location),
            (_, _, Some(location)) => start.merge(location),
            (None, None, None) => start,
        }
    }

    fn merge_location_without_values(
        start: aml_core::Location,
        attributes: &Attributes,
        children: &[AstNode],
    ) -> aml_core::Location {
        let children_location = children.iter().last().map(|node| node.location());

        match (children_location, attributes.location) {
            (Some(location), _) => start.merge(location),
            (_, Some(location)) => start.merge(location),
            (None, None) => start,
        }
    }
}

pub struct Parser {
    scope_stack: Vec<usize>,
    tokens: Tokens,
    ast: Ast,
}

impl Parser {
    pub fn new(mut tokens: Tokens) -> Self {
        tokens.consume_newlines();

        Self {
            tokens,
            ast: Ast::default(),
            scope_stack: Vec::new(),
        }
    }

    pub fn parse(mut self) -> Ast {
        let base_indent = match self.tokens.peek().kind() {
            TokenKind::Indent(indent) => indent,
            _ => 0,
        };

        self.ast.nodes = self.parse_block(base_indent);
        self.ast
    }

    fn parse_block(&mut self, block_indent: usize) -> Vec<AstNode> {
        self.add_scope();

        let mut nodes = vec![];

        loop {
            self.tokens.consume_newlines();

            let current_indent = match self.tokens.peek().kind() {
                TokenKind::Indent(i) => {
                    self.tokens.consume();
                    i
                }
                _ => 0,
            };

            if current_indent < block_indent {
                break;
            }

            if current_indent > block_indent {
                break;
            }

            if self.tokens.peek().kind() == TokenKind::Eof {
                break;
            }

            nodes.push(self.parse_node(current_indent));
        }

        self.pop_scope();
        nodes
    }

    fn parse_node(&mut self, current_indent: usize) -> AstNode {
        let token = self.tokens.peek_skip_indent();
        match token.kind() {
            TokenKind::Element(element) => self.parse_element(element, current_indent),
            TokenKind::Container(_) => self.parse_container(current_indent),
            TokenKind::String(_) => self.parse_string(),
            TokenKind::Identifier(_) => self.parse_identifier(),
            TokenKind::Decl => self.parse_declaration(),
            TokenKind::Local => self.parse_declaration(),
            TokenKind::Global => self.parse_declaration(),
            TokenKind::Component => self.parse_component(),
            TokenKind::ComponentSlot => self.parse_component_slot(),
            TokenKind::For => self.parse_for_loop(current_indent),
            TokenKind::If => self.parse_if_statement(current_indent),
            TokenKind::Switch => self.parse_switch_statement(current_indent),
            TokenKind::With => todo!(),
            _ => AstNode::Error(ErrorNode {
                token: token.kind(),
                location: token.location(),
            }),
        }
    }

    fn parse_element(&mut self, element: Element, current_indent: usize) -> AstNode {
        match element {
            Element::Text => self.parse_text(current_indent),
            Element::Span => self.parse_span(),
        }
    }

    fn parse_component_slot(&mut self) -> AstNode {
        let component = self.tokens.next_token();
        let start_location = component.location();
        let name = self.parse_identifier();

        let location = start_location.merge(name.location());
        AstNode::ComponentSlot(ComponentSlot {
            location,
            name: Box::new(name),
        })
    }

    fn parse_component(&mut self) -> AstNode {
        let component = self.tokens.next_token();
        let start_location = component.location();

        let name = self.parse_identifier();
        let attributes = self.maybe_parse_attributes();

        let location = match attributes.location {
            Some(location) => start_location.merge(location),
            None => start_location.merge(name.location()),
        };

        AstNode::Component(Component {
            name: Box::new(name),
            location,
            attributes,
        })
    }

    fn parse_text(&mut self, current_indent: usize) -> AstNode {
        let text = self.tokens.next_token();
        assert!(text.kind() == TokenKind::Element(Element::Text));

        let start_location = text.location();
        let attributes = self.maybe_parse_attributes();
        let values = self.parse_values();
        let children = self.maybe_parse_block(current_indent);

        let location = LocationCalculator::merge_location_with_values(
            start_location,
            &attributes,
            Some(&values),
            Some(&children),
        );

        AstNode::Text(Text {
            values,
            attributes,
            children,
            location,
            keyword: text.location(),
        })
    }

    fn parse_span(&mut self) -> AstNode {
        let span = self.tokens.next_token();
        assert!(span.kind() == TokenKind::Element(Element::Span));

        let start_location = span.location();
        let attributes = self.maybe_parse_attributes();
        let values = self.parse_values();

        let location =
            LocationCalculator::merge_location_without_values(start_location, &attributes, &values);

        AstNode::Span(Span {
            values,
            attributes,
            location,
            keyword: span.location(),
        })
    }

    fn parse_container(&mut self, current_indent: usize) -> AstNode {
        let token = self.tokens.next_token();
        let kind = token.kind().expect_container();
        let start_location = token.location();
        let attributes = self.maybe_parse_attributes();
        let children = self.maybe_parse_block(current_indent);

        let location = LocationCalculator::merge_location_without_values(
            start_location,
            &attributes,
            &children,
        );

        AstNode::Container(ContainerNode {
            kind,
            children,
            attributes,
            location,
            keyword: token.location(),
        })
    }

    fn parse_for_loop(&mut self, current_indent: usize) -> AstNode {
        let keyword = self.tokens.next_token();
        let start_location = keyword.location();
        self.tokens.consume_indent();

        let binding = self.parse_identifier();
        self.tokens.consume_indent();

        if self.tokens.peek().kind() == TokenKind::In {
            self.tokens.consume()
        }
        self.tokens.consume_indent();

        let value = parse_expression(&mut self.tokens);
        let children = self.maybe_parse_block(current_indent);

        let last_child_location = children.last().map(|node| node.location());

        let location = match (last_child_location, value.location()) {
            (Some(location), _) => start_location.merge(location),
            (_, location) => start_location.merge(location),
        };

        AstNode::For(For {
            binding: Box::new(binding),
            value,
            children,
            location,
            keyword: keyword.location(),
        })
    }

    fn parse_if_statement(&mut self, current_indent: usize) -> AstNode {
        let mut branches = vec![];
        let mut start_location = self.tokens.peek().location();

        while let Some(branch) = self.parse_conditional_branch(current_indent) {
            branches.push(branch);
        }

        let last_branch_location = branches.last().map(|branch| branch.location());
        let location = start_location.merge(last_branch_location.unwrap_or(start_location));

        AstNode::If(IfChain { branches, location })
    }

    fn parse_conditional_branch(&mut self, current_indent: usize) -> Option<ConditionalBranch> {
        match self.tokens.peek().kind() {
            TokenKind::If => Some(ConditionalBranch::If(self.parse_if_node(current_indent))),
            TokenKind::Else => {
                let next_token = self.tokens.next_token();
                let keyword = next_token.location();
                self.tokens.consume_indent();
                assert!(next_token.kind() == TokenKind::Else);

                match self.tokens.peek().kind() {
                    TokenKind::If => {
                        let if_node = self.parse_if_node(current_indent);
                        Some(ConditionalBranch::ElseIf(keyword, if_node))
                    }
                    _ => {
                        let children = self.maybe_parse_block(current_indent);
                        let last_child_location = children.last().map(|node| node.location());
                        let location = keyword.merge(last_child_location.unwrap_or(keyword));

                        Some(ConditionalBranch::Else(Else {
                            children,
                            location,
                            keyword,
                        }))
                    }
                }
            }
            _ => None,
        }
    }

    fn parse_if_node(&mut self, current_indent: usize) -> If {
        let keyword = self.tokens.next_token().location();
        self.tokens.consume_indent();
        let condition = parse_expression(&mut self.tokens);
        let then = self.maybe_parse_block(current_indent);

        let then_location = then.last().map(|node| node.location());
        let condition_location = condition.location();
        let location = keyword.merge(then_location.unwrap_or(condition_location));

        If {
            then,
            keyword,
            location,
            condition,
        }
    }

    fn parse_values(&mut self) -> Vec<AstNode> {
        let mut values = vec![];
        loop {
            let next_token = self.tokens.peek_skip_indent();
            match next_token.kind() {
                TokenKind::Newline => break,
                TokenKind::Eof => break,
                TokenKind::Identifier(_) => values.push(self.parse_identifier()),
                TokenKind::Primitive(_) => values.push(self.parse_primitive()),
                TokenKind::String(_) => values.push(self.parse_string()),
                token => {
                    self.tokens.consume();
                    values.push(AstNode::Error(ErrorNode {
                        token,
                        location: next_token.location(),
                    }))
                }
            }
        }
        values
    }

    fn parse_switch_statement(&mut self, current_indent: usize) -> AstNode {
        let keyword = self.tokens.next_token();
        let start_location = keyword.location();
        self.tokens.consume_indent();
        let condition = parse_expression(&mut self.tokens);

        let mut branches = vec![];
        while let Some(branch) = self.parse_switch_branch(current_indent) {
            branches.push(branch);
        }

        let last_branch_location = branches.last().map(|branch| branch.location());
        let location = start_location.merge(last_branch_location.unwrap_or(condition.location()));
        AstNode::Switch(SwitchChain { branches, location })
    }

    fn parse_switch_branch(&mut self, current_indent: usize) -> Option<SwitchBranch> {
        self.tokens.consume_newlines();

        match self.tokens.peek_skip_indent().kind() {
            TokenKind::Case => Some(self.parse_case_branch(current_indent)),
            TokenKind::Default => Some(self.parse_default_branch(current_indent)),
            _ => None,
        }
    }

    fn parse_case_branch(&mut self, current_indent: usize) -> SwitchBranch {
        let keyword = self.tokens.next_token().location();
        self.tokens.consume_indent();

        let condition = parse_expression(&mut self.tokens);
        self.tokens.consume_indent();

        let colon_token = self.parse_optional_colon();
        self.tokens.consume_indent();

        let children = self.parse_switch_children(current_indent);
        let location = self.calculate_switch_location(
            keyword,
            &children,
            colon_token,
            Some(condition.location()),
        );

        SwitchBranch::Case(SwitchCase {
            condition,
            children,
            location,
            keyword,
            has_colon: colon_token.is_some(),
        })
    }

    fn parse_default_branch(&mut self, current_indent: usize) -> SwitchBranch {
        let keyword = self.tokens.next_token().location();
        self.tokens.consume_indent();

        let colon_token = self.parse_optional_colon();
        self.tokens.consume_indent();

        let children = self.parse_switch_children(current_indent);
        let location = self.calculate_switch_location(keyword, &children, colon_token, None);

        SwitchBranch::Default(SwitchDefault {
            children,
            keyword,
            location,
            has_colon: colon_token.is_some(),
        })
    }

    fn parse_optional_colon(&mut self) -> Option<Token> {
        match self.tokens.peek().kind() {
            TokenKind::Operator(Operator::Colon) => Some(self.tokens.next_token()),
            _ => None,
        }
    }

    fn parse_switch_children(&mut self, current_indent: usize) -> Vec<AstNode> {
        match self.tokens.peek_skip_indent().kind() {
            kind if kind.is_declaration() => vec![self.parse_declaration()],
            TokenKind::Newline => self.maybe_parse_block(current_indent),
            TokenKind::Element(element) => {
                vec![self.parse_element(element, current_indent)]
            }
            _ => vec![],
        }
    }

    fn calculate_switch_location(
        &self,
        keyword: aml_core::Location,
        children: &[AstNode],
        colon_token: Option<aml_token::Token>,
        fallback_location: Option<aml_core::Location>,
    ) -> aml_core::Location {
        let last_child_location = children.last().map(|node| node.location());

        match (last_child_location, colon_token, fallback_location) {
            (Some(location), _, _) => keyword.merge(location),
            (_, Some(colon), _) => keyword.merge(colon.location()),
            (None, None, Some(fallback)) => keyword.merge(fallback),
            (None, None, None) => keyword,
        }
    }

    fn parse_primitive(&mut self) -> AstNode {
        let primitive = self.tokens.next_token();
        let TokenKind::Primitive(value) = primitive.kind() else { unreachable!() };
        let location = primitive.location();
        AstNode::Primitive(PrimitiveNode { location, value })
    }

    fn parse_string(&mut self) -> AstNode {
        let token = self.tokens.next_token();
        let location = token.location();
        AstNode::String(location)
    }

    fn parse_identifier(&mut self) -> AstNode {
        let token = self.tokens.next_token();
        let TokenKind::Identifier(location) = token.kind() else {
            return AstNode::Error(ErrorNode {
                location: token.location(),
                token: token.kind(),
            });
        };
        AstNode::Identifier(location)
    }

    fn parse_declaration(&mut self) -> AstNode {
        let keyword = self.tokens.next_token();
        let start_location = keyword.location();

        let kind = match keyword.kind() {
            TokenKind::Local => DeclarationKind::Local,
            TokenKind::Decl => DeclarationKind::Local,
            TokenKind::Global => DeclarationKind::Global,
            _ => unreachable!(),
        };

        self.tokens.consume_indent();
        let name = self.parse_identifier();

        self.tokens.consume_indent();
        self.tokens.consume(); // consume equal sign
        let value = parse_expression(&mut self.tokens);

        let location = start_location.merge(value.location());
        AstNode::Declaration(Declaration {
            kind,
            value,
            location,
            name: Box::new(name),
        })
    }

    fn maybe_parse_attributes(&mut self) -> Attributes {
        self.tokens.consume_indent();
        if self.tokens.peek_skip_indent().kind() == TokenKind::Operator(Operator::LBracket) {
            let attributes = self.parse_attributes();
            self.tokens.consume_indent();
            return attributes;
        }

        Attributes::default()
    }

    fn maybe_parse_block(&mut self, current_indent: usize) -> Vec<AstNode> {
        self.tokens.consume_newlines();

        let next_indent = match self.tokens.peek().kind() {
            TokenKind::Indent(i) => i,
            _ => 0,
        };

        match next_indent > current_indent {
            true => self.parse_block(next_indent),
            false => vec![],
        }
    }

    fn parse_attributes(&mut self) -> Attributes {
        let token = self.tokens.next_token();
        let start_location = token.location();
        assert!(token.kind() == TokenKind::Operator(Operator::LBracket));

        let mut attributes = vec![];
        let end_location = loop {
            let next_token = self.tokens.peek_skip_indent();

            match next_token.kind() {
                TokenKind::Operator(Operator::RBracket) => {
                    self.tokens.consume();
                    break next_token.location();
                }
                TokenKind::Eof => break next_token.location(),
                TokenKind::Newline => {
                    self.tokens.consume();
                    continue;
                }
                _ => {}
            }

            let name = self.parse_identifier();
            self.tokens.consume_all_whitespace();

            // TODO: this is a syntax error if there is no colon
            if self.tokens.peek().kind() == TokenKind::Operator(Operator::Colon) {
                self.tokens.consume();
            }

            self.tokens.consume_all_whitespace();
            let value = parse_expression(&mut self.tokens);

            if value.has_error() {
                loop {
                    match self.tokens.peek().kind() {
                        TokenKind::Operator(Operator::RBracket) => break,
                        TokenKind::Eof => break,
                        _ => self.tokens.consume(),
                    }
                }
            }

            let location = name.location().merge(value.location());
            attributes.push(AstNode::Attribute(Attribute {
                name: Box::new(name),
                value,
                location,
            }));

            self.skip_optional_comma();
        };

        Attributes {
            items: attributes,
            location: Some(start_location.merge(end_location)),
        }
    }

    fn skip_optional_comma(&mut self) {
        if self.tokens.peek_skip_indent().kind() == TokenKind::Operator(Operator::Comma) {
            self.tokens.consume();
        }
    }

    fn add_scope(&mut self) {
        let scope_id = self.ast.scopes.len();
        let parent = self.scope_stack.last().copied();

        self.ast.scopes.push(Scope {
            variables: Vec::new(),
            parent,
        });

        self.scope_stack.push(scope_id);
    }

    fn pop_scope(&mut self) {
        self.scope_stack.pop();
    }
}
