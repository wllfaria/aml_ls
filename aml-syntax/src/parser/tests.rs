use aml_token::{Lexer, Tokens};

use super::snapshots::SnapshotAst;
use crate::parser::Parser;

fn get_ast(template: &str) -> SnapshotAst<'_> {
    let tokens = Lexer::new(template).collect::<Vec<_>>();
    let tokens = Tokens::new(tokens, template.len());
    let parser = Parser::new(tokens);
    SnapshotAst::from_ast(parser.parse(), template)
}

#[test]
fn test_simple_text_element() {
    let template = r#"text "Hello""#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_text_element_with_attributes() {
    let template = r#"text [foreground: #ff0000] "Hello""#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_text_element_with_multiple_attributes() {
    let template = r#"text [foreground: #ff0000, background: #00ff00] "Hello""#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_text_with_multiple_values() {
    let template = r#"text "Hello" world true 1"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_simple_span_element() {
    let template = r#"span "Hello""#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_span_element_with_attributes() {
    let template = r#"span [foreground: #ff0000] "Hello""#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_span_element_with_multiple_attributes() {
    let template = r#"span [foreground: #ff0000, background: #00ff00] "Hello""#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_span_with_multiple_values() {
    let template = r#"span "Hello" world true 1"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_vstack_with_no_children() {
    let template = r#"vstack"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_vstack_with_children() {
    let template = r#"
vstack
    text "Hello"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_vstack_with_multiple_children_and_attributes() {
    let template = r#"
vstack [width: 10]
    text "Hello"
    text "World"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_hstack_with_children() {
    let template = r#"
hstack
    text "Hello"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_hstack_with_multiple_children_and_attributes() {
    let template = r#"
hstack [width: 10]
    text "Hello"
    text "World"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_border_with_no_children() {
    let template = r#"border"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_component() {
    let template = r#"@name"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_component_with_attributes() {
    let template = r#"@name [foreground: #ff0000]"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_component_slot() {
    let template = r#"$children"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_for_loop_inline_list() {
    let template = r#"
for item in [1, 2, 3]
    text item
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_for_loop_identifier() {
    let template = r#"
for item in list_of_items
    text item
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_for_loop_no_value() {
    let template = r#"
for item in
    text item
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_for_loop_no_binding() {
    let template = r#"
for  in [1,2,3]
    text item
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_function_call_on_variable() {
    let template = r#"
local result = func()
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_if_statement() {
    let template = r#"
if condition
    text "Hello"
else
    text "World"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_if_elseif_statement() {
    let template = r#"
if condition
    text "Hello"
else if condition
    text "World"
else
    text "!"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn test_malformed_if_elseif_statement() {
    let template = r#"
if condition
    text "Hello"
else if
    text "World"
else
    text "!"
"#;
    let ast = get_ast(template);
    insta::assert_yaml_snapshot!(ast);
}
