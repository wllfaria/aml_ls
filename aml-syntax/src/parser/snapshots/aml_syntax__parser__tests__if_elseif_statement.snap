---
source: aml-syntax/src/parser/tests.rs
expression: ast
---
nodes:
  - If:
      branches:
        - If:
            condition:
              Ident:
                value: condition
                location:
                  start_byte: 4
                  end_byte: 13
            then:
              - Text:
                  values:
                    - String:
                        value: "\"Hello\""
                        location:
                          start_byte: 23
                          end_byte: 30
                  attributes: []
                  children: []
                  text: "text \"Hello\""
                  location:
                    start_byte: 18
                    end_byte: 30
                  keyword:
                    start_byte: 18
                    end_byte: 22
            keyword:
              start_byte: 1
              end_byte: 3
            location:
              start_byte: 1
              end_byte: 30
            original: "if condition\n    text \"Hello\""
        - ElseIf:
            location:
              start_byte: 31
              end_byte: 35
            original: else
            if_node:
              condition:
                Ident:
                  value: condition
                  location:
                    start_byte: 39
                    end_byte: 48
              then:
                - Text:
                    values:
                      - String:
                          value: "\"World\""
                          location:
                            start_byte: 58
                            end_byte: 65
                    attributes: []
                    children: []
                    text: "text \"World\""
                    location:
                      start_byte: 53
                      end_byte: 65
                    keyword:
                      start_byte: 53
                      end_byte: 57
              keyword:
                start_byte: 36
                end_byte: 38
              location:
                start_byte: 36
                end_byte: 65
              original: "if condition\n    text \"World\""
        - Else:
            children:
              - Text:
                  values:
                    - String:
                        value: "\"!\""
                        location:
                          start_byte: 80
                          end_byte: 83
                  attributes: []
                  children: []
                  text: "text \"!\""
                  location:
                    start_byte: 75
                    end_byte: 83
                  keyword:
                    start_byte: 75
                    end_byte: 79
            location:
              start_byte: 66
              end_byte: 83
            keyword:
              start_byte: 66
              end_byte: 70
            original: "else\n    text \"!\""
      location:
        start_byte: 1
        end_byte: 83
      original: "if condition\n    text \"Hello\"\nelse if condition\n    text \"World\"\nelse\n    text \"!\""
variables: {}
scopes:
  - variables: []
    parent: ~
  - variables: []
    parent: 0
  - variables: []
    parent: 0
  - variables: []
    parent: 0
